src/osd/PrimaryLogPG.cc

```diff
@@ -9924,6 +9924,60 @@ void PrimaryLogPG::cancel_copy_ops(bool requeue, vector<ceph_tid_t> *tids)
   }
 }
 
+struct C_gather : public Context {
+  PrimaryLogPGRef pg;
+  hobject_t oid;
+  PrimaryLogPG::OpContext *ctx;
+  C_gather(PrimaryLogPG *pg_, hobject_t oid_, PrimaryLogPG::OpContext *ctx_) : pg(pg_), oid(oid_), ctx(ctx_) {}
+  void finish(int r) override {
+    if (r == -ECANCELED)
+      return;
+    // TODO: check errors
+    pg->lock();
+    pg->cls_gather_ops.erase(oid);
```

Comment:
I think here you need to check whether the op is still in cls_gather_ops.  If not, the op was necessarily cancelled and you should drop the lock and bail.  objecter cancellation is meant to free resources, and won't necessarily prevent that race.